packages/backend-dev-utils/src/ipcClient.ts

```diff
@@ -48,21 +50,40 @@ const IPC_TIMEOUT_MS = 5000;
 export class BackstageIpcClient {
   #messageId = 0;
   #sendMessage: SendMessage;
+  #handlers: Map<number, ResponseHandler> = new Map();
 
   /**
    * Creates a new client if we're in a child process with IPC and BACKSTAGE_CLI_CHANNEL is set.
    */
   static create(): BackstageIpcClient | undefined {
     const sendMessage = process.send?.bind(process);
-    return sendMessage && process.env.BACKSTAGE_CLI_CHANNEL
-      ? new BackstageIpcClient(sendMessage)
-      : undefined;
+    const client =
+      sendMessage && process.env.BACKSTAGE_CLI_CHANNEL
+        ? new BackstageIpcClient(sendMessage)
+        : undefined;
+
+    if (client) {
+      process.on('message', (message, _sendHandle) =>
+        client.handleMessage(message, _sendHandle),
+      );
+    }
+
+    return client;
   }
 
   constructor(sendMessage: SendMessage) {
     this.#sendMessage = sendMessage;
   }
 
+  private handleMessage(message: unknown, _sendHandle: unknown) {
+    const isResponse = (msg: unknown): msg is Response =>
+      (msg as Response)?.type === responseType;
+
+    if (isResponse(message)) {
+      this.#handlers.get(message.id)?.(message);
+    }
```

Comment:
```suggestion
    if ('type' in message && message.type === responseType) {
      const response = message as Response;
      this.#handlers.get(response.id)?.(response);
    }
```

Something like this? This is also what I'm thinking could be inlined in the listener too. There's not big difference between a cast and a type guard from a safety PoV afaik

Reply:
`in` operator does not work with unknown types.

Reply:
ah yep needs a truthiness and object type check too, maybe just `(message as Response)?.type === responseType` is fine

Reply:
Nope, that again requires type check 😄

Reply:
How so?

Reply:
Because if we do this: 
```ts
    if ((message as Response)?.type === responseType) {
      this.#handlers.get(message.id)?.(message);
    }
```
`tsc` will complain in the if-block:

```
packages/backend-dev-utils/src/ipcClient.ts:83:26 - error TS18046: 'message' is of type 'unknown'.

83       this.#handlers.get(message.id)?.(message);
                            ~~~~~~~

packages/backend-dev-utils/src/ipcClient.ts:83:40 - error TS2345: Argument of type 'unknown' is not assignable to parameter of type 'Response'.

83       this.#handlers.get(message.id)?.(message);
                                          ~~~~~~~
```

which of course requires casting again either by

```ts
this.#handlers.get((message as Response).id)?.(message as Response);
```

or 

```
const response = message as Response;
this.#handlers.get(response.id)?.(response);
```

and using the type guard makes sure it's `Response` inside the if-block as well. Don't know actually if type guard can be inlined in the `if` somehow but I bet not (?).

Reply:
It's all in one place now at least which I think is the main thing for readability